lib/graphql_metrics/tracer.rb

```diff
@@ -0,0 +1,125 @@
+# frozen_string_literal: true
+
+module GraphQLMetrics
+  class Tracer
+    # NOTE: Used to store timings from lexing, parsing, validation, before we have a context to store
+    # values in. Uses thread-safe Concurrent::ThreadLocalVar to store a set of values per thread.
+    cattr_accessor :pre_context
+
+    # NOTE: These constants come from the graphql ruby gem.
+    GRAPHQL_GEM_LEXING_KEY = 'lex'
+    GRAPHQL_GEM_PARSING_KEY = 'parse'
+    GRAPHQL_GEM_VALIDATION_KEYS = ['validate', 'analyze_query', 'analyze_multiplex']
+    GRAPHQL_GEM_TRACING_FIELD_KEYS = [
+      GRAPHQL_GEM_TRACING_FIELD_KEY = 'execute_field',
+      GRAPHQL_GEM_TRACING_LAZY_FIELD_KEY = 'execute_field_lazy'
+    ]
+
+    def self.trace(key, data)
+      # NOTE: Context doesn't exist yet during lexing, parsing.
+      possible_context = data[:query]&.context
+
+      skip_tracing = possible_context&.fetch(GraphQLMetrics::SKIP_GRAPHQL_METRICS_ANALYSIS, false)
+      return yield if skip_tracing
```

Comment:
After this early return I find the rest of the method a bit hard to follow. I wonder if it would read better as a big case statement the way the gem does? https://github.com/swalkinshaw/graphql-ruby/blob/1149e07bd8b5f64f4b1bd7265d95fdeb77d6baa5/lib/graphql/tracing/platform_tracing.rb#L21-L61

Reply:
Great idea!